crates/virtio-queue/src/descriptor_utils.rs

```diff
@@ -0,0 +1,939 @@
+    addr: Le64,
+    len: Le32,
+    flags: Le16,
+    next: Le16,
+}
+
+// Safe because it only has data and has no implicit padding.
+unsafe impl ByteValued for virtq_desc {}
+
+#[derive(Copy, Clone, Debug, Default)]
+#[repr(C)]
+struct virtq_avail {
+    flags: Le16,
+    idx: Le16,
+    ring: Le16,
+}
+
+// Safe because it only has data and has no implicit padding.
+unsafe impl ByteValued for virtq_avail {}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use crate::{Queue, QueueOwnedT, QueueT};
+    use vm_memory::{Bytes, GuestAddress};
+
+    const VIRTQ_DESC_F_NEXT: u16 = 0x1;
+    const VIRTQ_DESC_F_WRITE: u16 = 0x2;
+    const MAX_QUEUE_SIZE: u16 = 32768;
+
+    /// Test utility function to create a descriptor chain in guest memory.
+    pub fn create_descriptor_chain(
+        memory: &GuestMemoryMmap,
+        descriptor_array_addr: GuestAddress,
+        mut buffers_start_addr: GuestAddress,
+        descriptors: Vec<(DescriptorType, u32)>,
+        spaces_between_regions: u32,
+    ) -> Result<DescriptorChain<&GuestMemoryMmap>> {
+        let descriptors_len = descriptors.len();
+        for (index, (type_, size)) in descriptors.into_iter().enumerate() {
+            let mut flags = 0;
+            if let DescriptorType::Writable = type_ {
+                flags |= VIRTQ_DESC_F_WRITE;
+            }
+            if index + 1 < descriptors_len {
+                flags |= VIRTQ_DESC_F_NEXT;
+            }
+
+            let index = index as u16;
+            let desc = virtq_desc {
+                addr: buffers_start_addr.raw_value().into(),
+                len: size.into(),
+                flags: flags.into(),
+                next: (index + 1).into(),
+            };
+
+            let offset = size + spaces_between_regions;
+            buffers_start_addr = buffers_start_addr
+                .checked_add(u64::from(offset))
+                .ok_or(Error::InvalidChain)?;
+
+            let _ = memory.write_obj(
+                desc,
+                descriptor_array_addr
+                    .checked_add(u64::from(index) * std::mem::size_of::<virtq_desc>() as u64)
+                    .ok_or(Error::InvalidChain)?,
+            );
+        }
+
+        let avail_ring = descriptor_array_addr
+            .checked_add(
+                u64::from(descriptors_len as u16) * std::mem::size_of::<virtq_desc>() as u64,
+            )
+            .ok_or(Error::InvalidChain)?;
+        let avail = virtq_avail {
+            flags: 0.into(),
+            idx: 1.into(),
+            ring: 0.into(),
+        };
+        let _ = memory.write_obj(avail, avail_ring);
+
+        let mut queue: Queue = Queue::new(MAX_QUEUE_SIZE).unwrap();
+        queue
+            .try_set_desc_table_address(descriptor_array_addr)
+            .unwrap();
+        queue.try_set_avail_ring_address(avail_ring).unwrap();
+        queue.set_ready(true);
+        let desc = queue.iter(memory).unwrap().next().unwrap();
+        Ok(desc.clone())
+    }
```

Comment:
This feels like it duplicates a lot of what is available at https://github.com/rust-vmm/vm-virtio/blob/d739c5e/crates/virtio-queue/src/mock.rs#L445 (also see the other mock functions there). Is there something that this does fundamentally different?

Reply:
No, there is not. I could rewrite `create_descriptor_chain()` by using `build_desc_chain()`.

Reply:
I fixed the tests by using `build_desc_chain()`.